snapcraft/plugins/jhbuild.py

```diff
@@ -0,0 +1,383 @@
+        # Inform Snapcraft of the properties associated with pulling. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return [
+            'modules',
+            'module-set',
+            'module-set-dir',
+            'jhbuild-archive',
+            'jhbuild-mirror',
+        ]
+
+    @classmethod
+    def get_build_properties(cls):
+        # Inform Snapcraft of the properties associated with building. If these
+        # change in the YAML Snapcraft will consider the build step dirty.
+        return [
+            'cflags',
+        ]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.modules = [
+            module for module in self.options.modules if module[0] != '-']
+
+        self.skip_modules = [
+            module[1:] for module in self.options.modules if module[0] == '-']
+
+        self.build_packages += list(BUILD_PACKAGES)
+
+        self.jhbuild_src = os.path.join(self.partdir, 'jhbuild', 'src')
+        self.jhbuild_program = os.path.join(
+            self.partdir, 'jhbuild', 'usr', 'bin', 'jhbuild')
+        self.jhbuildrc_path = os.path.join(self.partdir, 'jhbuildrc')
+
+    def enable_cross_compilation(self):
+        """
+        We don't support cross compilation (yet?)
+        """
+        raise NotImplementedError(
+            'The {!s} plugin does not support cross compilation'.format(
+                self.name))
+
+    def run(self, cmd, cwd=None, **kwargs):
+        """Run a command.
+
+        Use run_output() if you need to capture or suppress output.
+
+        :param list cmd: command arguments, first is the executable
+        :param str cwd: working directory
+        """
+        return super().run(cmd, cwd=cwd, **kwargs)
+
+    def run_output(self, cmd, cwd=None, **kwargs):
+        """Run a command, capturing its output.
+
+        Use run() if you don't need the output and prefer it to be piped to
+        stdout.
+
+        :param list cmd: command arguments, first is the executable
+        :param str cwd: working directory
+        :return: the output of the command
+        :rtype: str
+        """
+        return super().run_output(cmd, cwd=cwd, **kwargs)
+
+    def jhbuild(self, args, output=True, **kwargs):
+        """Run JHBuild in the build stage.
+
+        :param list args: command arguments without executable
+        :return: the output of the command if captured
+        :rtype: str
+        """
+
+        cwd = kwargs.pop('cwd', os.getcwd())
+        cmd = []
+        if os.getuid() == 0:
+            cmd = ['sudo', '-H', '-u', 'jhbuild', '--']
+
+        cmd = cmd + [self.jhbuild_program,
+                     '--no-interact', '-f', self.jhbuildrc_path]
+
+        run = self.run_output if output else self.run
+        return run(cmd + args, cwd=cwd, **kwargs)
+
+    def _create_jhbuild_user(self):
+        common.run(['adduser', '--shell', '/bin/bash', '--disabled-password',
```

Comment:
Does this potentially require sudo, then? Might be worth mentioning in the docs at the top. In fact, it seems like a run may or may not fail depending on whether or not it was run with sudo and whether or not that user already exists. Should we instead _use_ sudo, here, thus prompting for a password if necessary?

I don't like creating users under the covers like this, by the way, but I assume it's required for jhbuild.